Space suits have a wear delay

Spookerton opened the pull request:
:cl: Jux, Spookerton
tweak: Space suits take a flat 5 seconds to put on.
bugfix: User-unique progress bars correctly indicate what you stopped interacting with when interrupted.
bugfix: Same-hand progress bars correctly check for same hand-ness when started right-handed.
/:cl:

closes #31196
other changes are necessary to make that PR viable so I put them together
this could be more extensive (adding a remove_delay, skills, etc) but I'm not doing that now

\- you can set equip_delay on items to make them take time to put in a slot
\- do_flags can be set with equip_delay_flags
\- behavior per-item (messages etc) can be added by overriding equip_delay_before, equip_delay_after
\- equip procs can be configured to skip the delay for round join, scene dolls, etc

SierraKomodo commented on code/modules/mob/inventory.dm:
```diff
@@ -1,97 +1,86 @@
+			do_flags ^= DO_FAIL_FEEDBACK
+		if (!do_after(src, I.equip_delay, I, do_flags))
+			return
+	equip_to_slot(I, slot, equip_flags & TRYEQUIP_REDRAW)
```
You're providing 3 parameters here, but the base proc definition only has 2 parameters

Spookerton replied:
The human override has three (which is what this ends up at), and the third will be ignored for non-humans.

Spookerton replied:
ie, this is existing behavior ([inventory.dm#L229](/Baystation12/Baystation12/tree/dev/code/modules/mob/living/carbon/human/inventory.dm#L229))

SierraKomodo reviewed (dismissed):
- The 'putting on suit' timer isn't visible to the person putting on the suit and there's no feedback that you're actually trying to put it on.
- Attempting to equip the suit multiple times causes multiple timers to proc that don't reset/cancel the old timers, which made for some odd behaviour.

SierraKomodo commented:
The runtimes were apparently a side effect of a bug in the version of byond I had installed and isn't relevant. The other two items are still applicable.

juxtaposed9930 commented:
> * The 'putting on suit' timer isn't visible to the person putting on the suit and there's no feedback that you're actually trying to put it on.
The bar should appear over the suit while it sits in the hand slot

SierraKomodo commented:
> 
> 
> > * The 'putting on suit' timer isn't visible to the person putting on the suit and there's no feedback that you're actually trying to put it on.
> 
> The bar should appear over the suit while it sits in the hand slot

Triple checked just to be sure - The progress bar isn't visible, and there's no chat message indicating you're trying to put the suit on.

That said I noticed a fun oddity - If I have a Turf tab open with myself on it, the progress bar appears _on that sprite_ in the turf tab, but not in word or over the item in hand.

Spookerton commented:
> That said I noticed a fun oddity - If I have a Turf tab open with myself on it, the progress bar appears _on that sprite_ in the turf tab, but not in word or over the item in hand.

That'll be an artifact of the do_after being public. I'd hoped it would be turf visible with that, since private *does* put it on the item in-hand and isn't a good indication of an action being taken. I'll pop it back to private if that's how it turned out, and add messages.

Spookerton commented:
Hm. You know, I don't like that. Not flexy enough.

Spookerton commented on code/_helpers/mobs.dm:
```diff
@@ -191,7 +191,7 @@ proc/age2agedescription(age)
 			. = DO_TARGET_CAN_TURN
 			break
-		if (!isnull(user_hand) && user_hand != user.hand)
+		if ((do_flags & DO_USER_SAME_HAND) && user_hand != user.hand)
```
user.hand is null when right, so this was skipping same-hand checks when started right-handed.

Spookerton commented on code/_helpers/mobs.dm:
```diff
@@ -218,7 +218,7 @@ proc/age2agedescription(age)
 				to_chat(user, SPAN_WARNING("You must remain on the same active hand to perform that action!"))
 			if (DO_USER_UNIQUE_ACT)
-				to_chat(user, SPAN_WARNING("You stop what you're doing with \the [user.do_unique_user_handle]."))
+				to_chat(user, SPAN_WARNING("You stop what you're doing with \the [target]."))
```
do_unique_user_handle is a sequential id, not the target

Spookerton commented:
Aight. Happy enough with this now.

SierraKomodo reviewed (dismissed):
Squash

Spookerton committed:
adding item equip delay, some do_after fixes

fixes do_after posting a bad message when halting for user unique constraint

fixes do_after failing to check hand swapping when starting on the right hand

refactors mob equipping to use a flags field instead of individual vars

adds item/var/equip_delay, item/var/equip_delay_flags, item/proc/equip_delay_before, item/proc/equip_delay_after
- uses equip_delay to do_after in equip_to_slot_if_possible when flags do not skip it
- equip_delay_flags are passed to do_after as flags
- equip_delay_before/after are virtual behavior functions items implementing equip_delay can override for messages etc

uses equip_delay for space suits, excluding hard/rig

SierraKomodo reviewed (approved):
Sure is fun when I don't get a GitHub notification for things.
